armi/runLog.py

```diff
@@ -98,82 +108,79 @@ class Log:
     Abstract class that ARMI code calls to be rendered to some kind of log.
     """
 
-    def __init__(self, verbosity=50):  # pylint: disable=unused-argument
+    def __init__(self, mpiRank=0, mpiSize=1):
         """
         Build a log object
 
         Parameters
         ----------
-        verbosity : int
-            if a msg verbosity is > this, it will be  emitted.
-            The default of 50 means only error messages will be emitted.
-            This usually gets adjusted by user settings quickly after instantiation.
+        mpiRank : int
+            If this is zero, we are in the parent process, otherwise child process.
+            The default of 0 means we assume the parent process.
+            This should not be adjusted after instantiation.
+        mpiSize : int
+            If this is one, we are in the parent process, with child process.
+            This should not be adjusted after instantiation.
 
         """
-        self._verbosity = verbosity
+        self._mpiRank = mpiRank
+        self._mpiSize = mpiSize
+        self._verbosity = logging.INFO
+        self.name = "log"
         self._singleMessageCounts = collections.defaultdict(lambda: 0)
         self._singleWarningMessageCounts = collections.defaultdict(lambda: 0)
-        self._outputStream = None
+        self.loggerName = (
+            "parent" if self._mpiRank == 0 else "{:>03d}".format(self._mpiRank)
+        )
+        self.logger = logging.getLogger(self.loggerName)
```

Comment:
I've been wanting this feature to come in for a while and I'm very happy about this PR 🤩 

I do have one comment that I'd like to raise here: Embedding ARMI as the "root" logger name. The reason is two-fold

1. This allows other applications to grab the logger (if they so choose) with `logger.getLogger("ARMI")`
2. This embeds the string `ARMI` in the log message if someone tweaks their log config, e.g.,
```python
>>> import logging
>>> logging.basicConfig(level=logging.INFO, format="%(name)s::%(message)s")
>>> lg = logging.getLogger("test-app")
>>> lg.info("Hello world")
test-app::Hello world
```

It might be useful to have the application logger "name" for the simply reflect the solver application (e.g., `DIF3D` vs. `ARMI.physics.neutronics.dif3d`). Might be arguments for both (grepping for all messages that relate to physics vs. reporting, bookkeeping, etc.). The "dot separated" structure is intentional as it build a hierarchy of loggers

> Loggers that are further down in the hierarchical list are children of loggers higher up in the list. For example, given a logger with a name of foo, loggers with names of foo.bar, foo.bar.baz, and foo.bam are all descendants of foo. 

- https://docs.python.org/3/library/logging.html#logger-objects

and for `Logger.propagate` - https://docs.python.org/3/library/logging.html#logging.Logger.propagate

> If this attribute evaluates to true, events logged to this logger will be passed to the handlers of higher level (ancestor) loggers, in addition to any handlers attached to this logger. Messages are passed directly to the ancestor loggers’ handlers - neither the level nor filters of the ancestor loggers in question are considered.

If the goal is to have solver-specific loggers, an "approved" way to grab a logger would be great. An interface could do something like
```python
from armi.runLog import buildChildLog

d3dLog = buildChildLog("DIF3D")   
# or
d3dLog = buildChildLog("DIF3D", "physics.neutronics") # to log as armi.physics.neutronics.DIF3D
```

Kind of a rambling mess but I hope something useful can be found 😅

Reply:
1. I like `"ARMI"` as a better default than `"parent"`, sure.  That makes sense to me.
2. As far as having things more custom to the module like `"DIF3D"` as the logger name, we love that. The idea is this is the first PR to refactor `runLog.py` to use the `logging` module. Then the _next_ PR would refactor the logging more to allow for various module-specific logging options. The primary one being module-by-module logging levels. But I like the idea of each logger having their own name. (I might have to play with that to make sure it's still convenient with MPI runs a bit.)

I like both of those ideas a lot. I can go change the default logger name, unless someone objects.

Reply:
Item 1 complete

Reply:
@youngmit  @ntouran   Okay, the latest version of the PR allows for people to have module-specific loggers.  Please see `armi/reactor/reactors.py` for an example of the usage.